--- charmrun.c	2010-01-26 16:48:12.000000000 -0600
+++ charmrun_timing.c	2010-01-28 19:13:16.000000000 -0600
@@ -2625,7 +2625,9 @@
 
 static void fast_idleFn(void) {sleep(0);}
 void finish_nodes(void);
-
+#define  getthetime(x) gettimeofday(&tim,NULL); x = tim.tv_sec + (tim.tv_usec/1000000.0);
+struct timeval tim;
+double t1,t2,t3;
 int main(int argc, char **argv, char **envp)
 {
   srand(time(0));
@@ -2649,6 +2651,7 @@
   nodetab_init();
 #endif
 /* Start the server port */
+  getthetime(t1);
   req_start_server();
   
   /* Initialize the IO module */
@@ -2721,7 +2724,8 @@
   kill_nodes();
 #endif
   if(arg_verbose) fprintf(stderr, "Charmrun> node programs all connected\n");
-
+ getthetime(t2);
+ printf("Total time  = %f\n" t2-t1);
   /* enter request-service mode */
 	//Uncomment me
   // while (1) req_poll();
